doc/SONiC_VRRP_HLD.md

```diff
@@ -0,0 +1,582 @@
+
+
+
+**Example:- Entery with multiple virtual IPs**
+
+admin@sonic:~$ redis-cli -n 4 HGETALL "VRRP|Vlan1|1"
+ 1) "vrid"
+ 2) "1"
+ 3) "vip"
+ 4) "4.1.1.100,4.1.1.200,4.1.1.50,4.1.1.150,"
+ 5) "priority"
+ 6) "80"
+ 7) "adv_interval"
+ 8) "1"
+ 9) "state"
+10) ""
+11) "version"
+12) "2"
+13) "pre_empt"
+14) "True"
+15) "track_interface"
+16) "Ethernet7|weight|10,PortChannel001|weight|10,"
+
+### 4.2.2 APP_DB Changes
+
+**VRRP_TABLE**
+
+Producer:  VrrpSyncd 
+
+Consumer: VrrpOrch
+
+Description: This is a new table that contains VRRP state information. This entry will be added to APP_DB for each VIP for every Master VRRP instance. 
+
+Schema:
+
+```
+; New table
+; holds the VRRP state and VMAC information
+
+key = VRRP_TABLE:interface_name:vip:type
+    interface_name        ; interface name as a string. Vlan, Ethernet or PortChannel
+    vip                   ; virtual IP address in a.b.c.d/32 format
+    type                  ; IP address type string. only IPv4 supported currently
+    
+; field = value
+vmac = virtual_mac_address ; Virtual MAC address associated with VRRP instance
+```
+
+Example:- 
+
+**Key**: VRRP_TABLE:Vlan1000:[40.10.8.101/32:ipv4]
+
+**Value**: "vmac":"00:00:5e:00:01:08"
+
+## 4.3 Modules Design and Flows
+
+Below diagram depicts interactions of submodules for VRRP support.
+
+```mermaid
+sequenceDiagram
+    participant config_db
+    participant vrrpmgrd
+    participant keepalived
+    participant kernel
+    participant vrrpsyncd
+    participant app_db 
+    participant vrrporchd
+    participant asic_db
+
+    config_db->>vrrpmgrd: VRRP configuration changes
+    vrrpmgrd->>keepalived: Update configuration
+    Note over keepalived: Run VRRP state machine
+    keepalived->>kernel: Add/Del MACVLAN interface with virtual IP
+    kernel->>vrrpsyncd: Add/Del IP to MACVLAN interface
+    Note over vrrpsyncd: Add to APP_DB for Master instance
+    vrrpsyncd->>app_db: Add/Del interface, virtual IP
+    app_db->>vrrporchd: Listen to VRRP master/standby state
+    vrrporchd->>asic_db: Add/Del VMAC for VRRP interface 
+    
+
+
+```
+
+### vrrpmgrd
+
+- Listens to VRRP create, delete and parameter change in config_db. 
+- Upon change,
+  - Updates keepalived.conf file with new configuration
+  - Sends a signal to keepalived to read and apply new configuration from keepalived.conf file.
+
+### keepalived
+
+- Reads the configuration from keepalived.conf file.
+- Listens to kernel link and IP events.
+- Runs VRRP protocol state machine, 
+  - For VRRP instances state transition to Master programs a MACVLAN interface with VIP as interface IP in kernel
+  - For VRRP instances state transition to non-Master deletes MACVLAN interface, if programed, from kernel.
+
+### vrrpsyncd
+
+Listens to MACVLAN interface programming in kernel. Status of MACVLAN interface determines Master/Backup state of VRRP instances. VRRP_Table in APP_DB will be programmed with interface name and VIP for Master instances.
```

Comment:
macvlan interface can be used for other use cases, are we assuming all macvlan interface are created by keepalived?

Reply:
No, only MACVLANs whose interface name starts with 'vrrp' in kernel is associated with VRRP. keepalived adds MACVLANs with name starting with 'vrrp' substring.